mathics/core/expression.py

```diff
@@ -682,26 +679,25 @@ def evaluate(
                 if hasattr(expr, "options") and expr.options:
                     evaluation.options = expr.options
 
-                # This calls evaluate_next. This routine implements a single
-                # step in the evaluation, and determines if a fixed point
-                # was reached (reevaluate->False).
-                # Notice that evaluate_next calls ``evaluate``
-                # for the other ``BaseElement`` subclasses.
+                # ``rewrite_apply_eval_step()`` makes a pass at
+                # evaluating the expression. If we know that a further
+                # evaluation will not be needed, ``reevaluate`` is set
+                # False.  Note that ``rewrite_apply_eval_step()`` can
+                # perform further ``evaluate`` and we will recurse
+                # back into this routine.
                 expr, reevaluate = expr.rewrite_apply_eval_step(evaluation)
 
                 if not reevaluate:
                     break
 
-                # Trace evaluation...
+                # TraceEvaluation[] logging.
                 if evaluation.definitions.trace_evaluation:
                     evaluation.print_out(
                         "  " * evaluation.recursion_depth + "-> %s" % expr
                     )
                 iteration += 1
-                # Check if the iterationlimit was reached.
-                # we need to check on each step, in case that the expression
-                # changes its value. Maybe there is another way, for example,
```

Comment:
I think that the way this is done is exactly right. I don't see anything suboptimal about this. 

Adding fields in the evaluation, even if it wasn't wrong (and I think it has a good chance of being wrong) would require extra work in maintaining.